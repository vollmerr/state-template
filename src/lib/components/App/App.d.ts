import * as React from 'react';
import * as Router from 'react-router-dom';

import * as types from '../../utils/types';

export interface AppProps {
  /** Render a custom header */
  renderHeader?: (props: AppProps) => React.ReactNode;

  /** Render custom routing */
  renderRouting?: (props: AppProps) => React.ReactNode;

  /** Render a custom footer */
  renderFooter?: (props: AppProps) => React.ReactNode;

  /** Fix header to top */
  fixedHeader?: boolean;

  /** Align header nav to right */
  alignHeader?: 'center'
    | 'left'
    | 'right';

  /** Custom header branding logo */
  brandingLogo?: types.BrandingLogo;

  /** Routes to render and build nav from */
  routes?: Array<types.Route>;

  /** Redirect route if no matching route */
  redirect?: string;

  /** Link to use for all 'contact us' links */
  contactLink?: types.ContactLink;

  /** react-router-dom router to use */
  router?: Router.BrowserRouter
    | Router.HashRouter
    | Router.MemoryRouter
    | Router.StaticRouter;
}

declare class App extends React.Component<AppProps, {}> {}

export default App;
