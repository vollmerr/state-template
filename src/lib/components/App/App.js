import React from 'react';
import T from 'prop-types';
import { HashRouter } from 'react-router-dom';
import classNames from 'classnames';
import debounce from 'lodash.debounce';

import * as propUtils from '../../utils/propTypes';
import ErrorBoundary from '../ErrorBoundary';
import Messages from '../Messages';

import Routing from '../Routing';
import Header from '../Header';
import Footer from '../Footer';

import '../../style/core/css/cagov.core.css';
import '../../style/style.scss';

/**
 * This container handles updating user setttings, rendering
 * the header and footer, and routing. This container is the
 * default export fhe project, and the main entry point of any
 * application that is using this project.
 */
export class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = { returnTopVisible: false };
    this.scrollEvent = debounce(this.onScroll, 200);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.scrollEvent);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scrollEvent);
  }

  // handles displaying the return to top button if scrolled
  onScroll = () => {
    const { returnTopVisible } = this.state;
    const isVisible = window.scrollY > 130;

    if (returnTopVisible !== isVisible) {
      this.setState({ returnTopVisible: isVisible });
    }
  }

  // scrolls window to top
  onClickReturnTop = () => {
    window.scroll(0, 0);
  }

  // render return to top button
  renderReturnTop = () => {
    const { returnTopVisible } = this.state;
    const cn = classNames([
      'return-top',
      { 'is-visible': returnTopVisible },
    ]);

    return (
      <button
        type={'button'}
        className={cn}
        onClick={this.onClickReturnTop}
      />
    );
  }

  // renders default header if no `renderHeader` passed
  renderHeader = () => {
    const {
      renderHeader,
      headerProps,
      routingProps,
    } = this.props;

    if (renderHeader) {
      return renderHeader(this.props);
    }

    return <Header {...routingProps} {...headerProps} />;
  }

  // renders default routing if no `renderRouting` passed
  renderRouting = () => {
    const {
      renderRouting,
      routingProps,
    } = this.props;

    if (renderRouting) {
      return renderRouting(this.props);
    }

    return <Routing {...routingProps} />;
  }

  // renders default footer if no `renderFooter` passed
  renderFooter = () => {
    const {
      renderFooter,
      footerProps,
      routingProps,
    } = this.props;

    if (renderFooter) {
      return renderFooter(this.props);
    }

    return <Footer {...routingProps} {...footerProps} />;
  }

  render() {
    const {
      router: Router,
    } = this.props;

    return (
      <Router>
        <ErrorBoundary>
          {this.renderHeader()}
          {this.renderRouting()}
          {this.renderFooter()}
          {this.renderReturnTop()}
          <Messages />
        </ErrorBoundary>
      </Router>
    );
  }
}

App.propTypes = {
  /** Render a custom header */
  renderHeader: T.func,
  /** Render custom routing */
  renderRouting: T.func,
  /** Render a custom footer */
  renderFooter: T.func,
  /** react-router-dom router type */
  router: T.func,
  /** Props to pass to header */
  headerProps: T.shape({
    fixed: T.bool,
    align: T.oneOf(['center', 'left', 'right']),
    brandingLogo: propUtils.brandingLogo,
  }),
  /** Props to pass to routing */
  routingProps: T.shape({
    routes: T.arrayOf(propUtils.route),
    redirect: T.string,
    contactLink: propUtils.contactLink,
  }),
  /** Props to pass to footer (current none) */
  footerProps: T.object,
};

App.defaultProps = {
  renderHeader: null,
  renderRouting: null,
  renderFooter: null,
  router: HashRouter,
  headerProps: {},
  routingProps: {},
  footerProps: {},
};

export default App;
