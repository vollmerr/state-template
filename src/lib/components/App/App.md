```jsx
import { createMemoryHistory } from 'history';

const Home = () => <div>in Home</div>;
const Help = () => <div>in Help</div>;

const routes = [
  {
    key: 'home',
    name: 'Home',
    path: '/',
    exact: true,
    icon: 'ca-gov-icon-home',
    component: Home,
  },
  {
    key: 'help',
    name: 'Help',
    path: '/help',
    exact: true,
    icon: 'ca-gov-icon-question-line',
    component: Help,
  },
];

const props = {
  fixedHeader: false,
  alignHeader: 'right',
  routes,
  contactLink: {
    text: 'Contact Link!',
    href: 'https://cdt.ca.gov/support/',
  },
  history: createMemoryHistory(),
};

<div style={{ overflow: 'auto', display: 'grid' }}>
  <App {...props} />
</div>
```

### custom header, routing, and footer

```jsx
import { createMemoryHistory } from 'history';

const renderHeader = (props) => <header>custom header</header>;
const renderRouting = (props) => <div>custom routing</div>;
const renderFooter = (props) => <footer>custom footer</footer>;

<App
  renderHeader={renderHeader}
  renderRouting={renderRouting}
  renderFooter={renderFooter}
  history={createMemoryHistory()}
/>
```
