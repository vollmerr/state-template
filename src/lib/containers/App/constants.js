export const UPDATE_DISPLAY = 'ST/App/UPDATE_DISPLAY';
