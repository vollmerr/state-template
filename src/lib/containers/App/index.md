
```jsx
  const { MemoryRouter } = require('react-router-dom');

  const Home = () => <div>in Home</div>;
  const Help = () => <div>in Help</div>;

  const routes = [
    {
      key: 'home',
      name: 'Home',
      path: '/',
      exact: true,
      icon: 'ca-gov-icon-home',
      component: Home,
    },
    {
      key: 'help',
      name: 'Help',
      path: '/help',
      exact: true,
      icon: 'ca-gov-icon-question-line',
      component: Help,
    },
  ];

  <div style={{ overflow: 'auto', display: 'grid' }}>
    <App routes={routes} router={MemoryRouter} />
  </div>
```
