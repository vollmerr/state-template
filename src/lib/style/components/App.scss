.return-top {
  border: 0;
}
