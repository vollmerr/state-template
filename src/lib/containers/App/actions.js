import { createAction } from 'redux-actions';

import * as C from './constants';

export const updateRouting = createAction(C.UPDATE_ROUTING);
export const updateDisplay = createAction(C.UPDATE_DISPLAY);
export const updateStatus = createAction(C.UPDATE_STATUS);
