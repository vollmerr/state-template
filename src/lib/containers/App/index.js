import React from 'react';
import T from 'prop-types';
import { connect } from 'react-redux';

import { routeProp } from '../../utils/propTypes';

import * as routerActions from '../Router/actions';
import Router from '../Router';
import Header from './Header';
import Footer from './Footer';

import odiLogo from '../../images/office-of-digital-innovation-logo.png';
import '../../style/core/css/cagov.core.css';
import '../../style/style.scss';

// main container for applications - handles routing and user settings
export class App extends React.Component {
  componentDidMount() {
    this.setSettings();
  }

  // set setttings user has set (in SettingsBar)
  setSettings = () => {
    document.body.style.zoom = Number(localStorage.getItem('zoom'));
    if (localStorage.getItem('highContrast')) {
      document.getElementsByTagName('html')[0].classList.add('high-contrast');
    }
  }

  render() {
    const { routes, brandingLogo } = this.props;

    return (
      <>
        <Header brandingLogo={brandingLogo} />
        <Router routes={routes} />
        <Footer />
      </>
    );
  }
}

App.propTypes = {
  routes: T.arrayOf(routeProp).isRequired,
  brandingLogo: T.string,
};

App.defaultProps = {
  brandingLogo: odiLogo,
};

export const mapDispatchToProps = dispatch => ({
  updateRouting: props => dispatch(routerActions.updateRouting(props)),
});

const withRedux = connect(null, mapDispatchToProps);

export default withRedux(App);
