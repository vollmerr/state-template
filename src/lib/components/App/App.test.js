import React from 'react';
import { shallow } from 'enzyme';

import App from './App';

const props = {
  history: { push: jest.fn() },
};

let wrapper;
let instance;
describe('App', () => {
  beforeEach(() => {
    wrapper = shallow(<App {...props} />);
    instance = wrapper.instance();
  });

  describe('onScroll', () => {
    it('should do nothing if scrolled up and return top is not visible', () => {
      window.scrollY = 0;
      instance.onScroll();
      expect(wrapper.state('returnTopVisible')).toBeFalsy();
    });

    it('should set return top as visible if scrolled up and return top is visible', () => {
      window.scrollY = 0;
      wrapper.setState({ returnTopVisible: true });
      instance.onScroll();
      expect(wrapper.state('returnTopVisible')).toBeFalsy();
    });

    it('should set return top as visible if scrolled down and return top is not visible', () => {
      window.scrollY = 99999;
      instance.onScroll();
      expect(wrapper.state('returnTopVisible')).toBeTruthy();
    });

    it('should do nothing if scrolled down and return top is visible', () => {
      window.scrollY = 99999;
      wrapper.setState({ returnTopVisible: true });
      instance.onScroll();
      expect(wrapper.state('returnTopVisible')).toBeTruthy();
    });
  });

  describe('renderReturnTop', () => {
    it('should render a component', () => {
      const ReturnTop = shallow(instance.renderReturnTop());
      expect(ReturnTop.children.length).not.toBe(0);
    });

    it('should attach the `is-visible` class if visible', () => {
      wrapper.setState({ returnTopVisible: true });
      const ReturnTop = shallow(instance.renderReturnTop());
      expect(ReturnTop.prop('className')).toMatch(/is-visible/);
    });

    it('should call `onClickReturnTop` when clicked', () => {
      instance.onClickReturnTop = jest.fn();
      const ReturnTop = shallow(instance.renderReturnTop());
      ReturnTop.prop('onClick')();
      expect(instance.onClickReturnTop).toBeCalled();
    });
  });

  describe('renderHeader', () => {
    it('should render the custom routing if `renderHeader` passed', () => {
      const renderHeader = () => <header>custom header</header>;
      wrapper.setProps({ renderHeader });
      const isSame = shallow(instance.renderHeader()).equals(renderHeader());
      expect(isSame).toBeTruthy();
    });
  });

  describe('renderRouting', () => {
    it('should render the custom routing if `renderRouting` passed', () => {
      const renderRouting = () => <div>custom routing</div>;
      wrapper.setProps({ renderRouting });
      const isSame = shallow(instance.renderRouting()).equals(renderRouting());
      expect(isSame).toBeTruthy();
    });
  });

  describe('renderFooter', () => {
    it('should render the custom footer if `renderFooter` passed', () => {
      const renderFooter = () => <footer>custom footer</footer>;
      wrapper.setProps({ renderFooter });
      const isSame = shallow(instance.renderFooter()).equals(renderFooter());
      expect(isSame).toBeTruthy();
    });
  });

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should allow a custom router to be used that is passed the history', () => {
      const CustomRouter = () => <div>in custom routing...</div>;
      wrapper.setProps({ router: CustomRouter });
      expect(wrapper.find(CustomRouter).length).toBe(1);
      expect(wrapper.find(CustomRouter).prop('history')).toBe(props.history);
    });
  });
});
