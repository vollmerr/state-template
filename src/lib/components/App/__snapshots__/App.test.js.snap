// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render should render correctly 1`] = `
<Router
  history={Object {}}
>
  <ErrorBoundary>
    <Header
      align="right"
      brandingLogo={
        Object {
          "alt": "Office of Digital Innovation logo",
          "href": "https://cdt.ca.gov/",
          "src": "office-of-digital-innovation-logo.png",
        }
      }
      contactLink={
        Object {
          "href": "https://cdt.ca.gov/support/",
          "text": "Contact Us",
        }
      }
      fixed={true}
      routes={Array []}
      title={null}
    />
    <withRouter(Routing)
      redirect={null}
      routes={Array []}
    />
    <Footer
      contactLink={
        Object {
          "href": "https://cdt.ca.gov/support/",
          "text": "Contact Us",
        }
      }
    />
    <button
      className="return-top"
      onClick={[Function]}
      type="button"
    />
  </ErrorBoundary>
</Router>
`;
